presto-main/src/test/java/io/prestosql/operator/TestGroupedTopNRankBuilder.java

```diff
@@ -0,0 +1,283 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.prestosql.operator;
+
+import com.google.common.collect.ImmutableList;
+import com.google.common.primitives.Ints;
+import io.prestosql.spi.Page;
+import io.prestosql.spi.type.Type;
+import io.prestosql.spi.type.TypeOperators;
+import io.prestosql.sql.gen.JoinCompiler;
+import io.prestosql.type.BlockTypeOperators;
+import org.testng.annotations.BeforeMethod;
+import org.testng.annotations.DataProvider;
+import org.testng.annotations.Test;
+
+import java.util.List;
+import java.util.Optional;
+import java.util.concurrent.atomic.AtomicBoolean;
+
+import static com.google.common.collect.Iterables.getOnlyElement;
+import static io.prestosql.RowPageBuilder.rowPageBuilder;
+import static io.prestosql.RowPagesBuilder.rowPagesBuilder;
+import static io.prestosql.operator.PageAssertions.assertPageEquals;
+import static io.prestosql.operator.UpdateMemory.NOOP;
+import static io.prestosql.spi.connector.SortOrder.ASC_NULLS_LAST;
+import static io.prestosql.spi.type.BigintType.BIGINT;
+import static io.prestosql.spi.type.DoubleType.DOUBLE;
+import static org.testng.Assert.assertEquals;
+import static org.testng.Assert.assertFalse;
+import static org.testng.Assert.assertTrue;
+
+public class TestGroupedTopNRankBuilder
```

Comment:
Make sure you test peer groups with `nulls` and `NaN`

Reply:
This actually wouldn't test GroupedTopNRankBuilder since GroupedTopNRankBuilder takes in the comparison and equality strategies as inputs out of its control. We'd be better off testing the whole query if that's the concern.

Reply:
Given that we use:
```
        typeOperators = new TypeOperators();
        blockTypeOperators = new BlockTypeOperators(typeOperators);
```
I think some basic test here would also work since it would use same comparators, equals as non-test code (just not compiled version)